### DISCUSSION TITLE
Validating decimal max_digits with an optional field

### DISCUSSION THREAD

**Discussion by piercsi** [pydantic/pydantic#8539] [Question]
I have a flat file where I need to validate string decimals from an external file; empty values are stored as 'NULL'.

How can I apply the max_digits and decimal_places validators so they only apply to non-NULL values?

The code below returns `ValueError: Unknown constraint max_digits`

```
from decimal import Decimal
from pydantic import BaseModel, BeforeValidator, Field
from typing import Annotated


class MyDecimalModel(BaseModel):
    my_decimal: Annotated[Decimal | None, BeforeValidator(lambda o: None if o == 'NULL' else Decimal(o)), Field(max_digits=10, decimal_places=5)]


assert MyDecimalModel(my_decimal='3.12').my_decimal == Decimal('3.12')
assert MyDecimalModel(my_decimal='NULL').my_decimal is None
```

(pydantic 2.5.3)

**Comment by pythonweb2** [comment]
See #6955, I ran into this recently as well.
